pkg/winapi/winapi.go

```diff
@@ -0,0 +1,245 @@
+	PeakVirtualSize              uintptr       // SIZE_T
+	VirtualSize                  uintptr       // SIZE_T
+	PageFaultCount               uint32        // ULONG
+	PeakWorkingSetSize           uintptr       // SIZE_T
+	WorkingSetSize               uintptr       // SIZE_T
+	QuotaPeakPagedPoolUsage      uintptr       // SIZE_T
+	QuotaPagedPoolUsage          uintptr       // SIZE_T
+	QuotaPeakNonPagedPoolUsage   uintptr       // SIZE_T
+	QuotaNonPagedPoolUsage       uintptr       // SIZE_T
+	PagefileUsage                uintptr       // SIZE_T
+	PeakPagefileUsage            uintptr       // SIZE_T
+	PrivatePageCount             uintptr       // SIZE_T
+	ReadOperationCount           int64         // LARGE_INTEGER
+	WriteOperationCount          int64         // LARGE_INTEGER
+	OtherOperationCount          int64         // LARGE_INTEGER
+	ReadTransferCount            int64         // LARGE_INTEGER
+	WriteTransferCount           int64         // LARGE_INTEGER
+	OtherTransferCount           int64         // LARGE_INTEGER
+}
+
+type kWaitReason int32
+
+type clientID struct {
+	UniqueProcess uintptr // HANDLE
+	UniqueThread  uintptr // HANDLE
+}
+
+type systemThreadInformation struct {
+	KernelTime      int64       // LARGE_INTEGER
+	UserTime        int64       // LARGE_INTEGER
+	CreateTime      int64       // LARGE_INTEGER
+	WaitTime        uint32      // ULONG
+	StartAddress    uintptr     // PVOID
+	ClientId        clientID    // CLIENT_ID
+	Priority        kPriority   // KPRIORITY
+	BasePriority    int32       // LONG
+	ContextSwitches uint32      // ULONG
+	ThreadState     uint32      // ULONG
+	WaitReason      kWaitReason // KWAIT_REASON
+}
+
+func add(p unsafe.Pointer, x uintptr) unsafe.Pointer {
+	return unsafe.Pointer(uintptr(p) + x)
+}
+
+func checkProcNtQuerySystemInformationAvailable() error {
+	if ntDLLLoadErr != nil {
+		return errors.Wrap(ntDLLLoadErr, "winapi: can't load dll Ntdll.dll")
+	}
+	if procNtQuerySystemInformationLoadErr != nil {
+		return errors.Wrap(procNtQuerySystemInformationLoadErr, "winapi: can't get procedure NtQuerySystemInformation")
+	}
+	return nil
+}
+
+func GetSystemProcessorPerformanceInformation() ([]SystemProcessorPerformanceInformation, error) {
+	if err := checkProcNtQuerySystemInformationAvailable(); err != nil {
+		return nil, err
+	}
+
+	// Make maxResults large for safety.
+	// We can't invoke the api call with a results array that's too small.
+	// If we have more than 2056 cores on a single host, then it's probably the future.
+	maxBuffer := 2056
+	// buffer for results from the windows proc
+	resultBuffer := make([]SystemProcessorPerformanceInformation, maxBuffer)
+	// size of the buffer in memory
+	bufferSize := uintptr(systemProcessorPerformanceInfoSize) * uintptr(maxBuffer)
+	// size of the returned response
+	var retSize uint32
+
+	retCode, _, err := procNtQuerySystemInformation.Call(
+		systemProcessorPerformanceInformationClass, // System Information Class -> SystemProcessorPerformanceInformation
```

Comment:
Windows docs suggest that you should no use `SystemProcessorPerformanceInformation` because it can differ from version to version and be discounted in the next version. They suggest using `GetSystemTimes` which is not per CPU :( 
https://docs.microsoft.com/en-us/windows/desktop/api/winternl/nf-winternl-ntquerysysteminformation#system_processor_performance_information

also: 
```
[NtQuerySystemInformation may be altered or unavailable in future versions of Windows. Applications should use the alternate functions listed in this topic.]
```

Reply:
Is it ok to leave with it? Should we test on the each major version first?

Reply:
> Is it ok to leave with it? Should we test on the each major version first?
Yeah, I've seen this warning regarding the usage of this winapi call.
In fact, any win api call can be altered in the future.

It's already tested on Win7,8,10 and WinServer 2012/2016/2019

Actually, the idea of using this call came from browsing signalfx/signalfx-agent implementation: https://github.com/signalfx/signalfx-agent/blob/e89f2c5fad499a2637010ac32a48b16a0d7b4173/internal/monitors/cpu/cpu_windows.go